plutus-use-cases/src/Language/Plutus/Coordination/Contracts/CrowdFunding.hs

```diff
@@ -0,0 +1,150 @@
+-- | Crowdfunding contract implemented using the [[Plutus]] interface.
+-- This is the fully parallel version that collects all contributions
+-- in a single transaction. This is, of course, limited by the maximum
+-- number of inputs a transaction can have.
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE RecordWildCards   #-}
+{-# OPTIONS -fplugin=Language.Plutus.CoreToPLC.Plugin #-}
+module Language.Plutus.Coordination.Contracts.CrowdFunding (
+    Campaign(..)
+    -- * Functionality for campaign contributors
+    , contribute
+    , contributionScript
+    , refund
+    , refundTrigger
+    -- * Functionality for campaign owners
+    , collect
+    , collectFundsTrigger
+    ) where
+
+import           Language.Plutus.Coordination.Plutus
+import           Language.Plutus.CoreToPLC.Plugin     (plc)
+import qualified Language.Plutus.CoreToPLC.Primitives as Prim
+
+import           Prelude                              (Bool (..), Maybe (..), Num (..), Ord (..), succ, sum, ($))
+
+-- | A crowdfunding campaign.
+data Campaign = Campaign
+    { campaignDeadline           :: !BlockHeight
+    , campaignTarget             :: !Value
+    , campaignCollectionDeadline :: !BlockHeight
+    , campaignOwner              :: !PubKey
+    }
+
+-- | Contribute funds to the campaign (contributor)
+--
+contribute :: Campaign -> Value -> TxM [TxOutRef]
+contribute c value = do
+    assert (value > 0)
+    contributorPubKey <- lookupMyPubKey
+    myPayment         <- createPayment (value + standardTxFee)
+    let validator = contributionScript c (Just contributorPubKey)
+        o = TxOutScript
+            value
+            (hash validator)
+            0 -- TODO: contributorPubKey ought to be lifted into PLC at coordination runtime as the data script
+    submitTransaction Tx
+      { txInputs  = (myPayment, myPayment) -- TODO: Change to [myPayment] when we can have a list of inputs
+      , txOutputs = (o, o)
+      }
+    -- the transaction above really ought to be merely a transaction *template* and the transaction fee ought to be
+    -- added by the Wallet API Plutus library on the basis of the size and other costs of the transaction
+
+-- | The validator script that determines whether the campaign owner can
+--   retrieve the funds or the contributors can claim a refund.
+--
+--   See note [Contracts and Validator Scripts] in
+--       Language.Plutus.Coordination.Contracts
+contributionScript ::
+       Campaign
+    -> Maybe PubKey
+    -> PlutusTx
+contributionScript _ _  = PlutusTx inner where
```

Comment:
then the public key is ignored? Perhaps this relates to your note and I'm just being dense...

Reply:
`Campaign` and `Maybe PubKey` show up as the last two arguments in the `inner` lambda. My plan is for the Plutus coordinating code to translate them to PLC at "script generation time" and then apply them to obtain a real validator script which expects three arguments (data, redeemer, txn info). I tried to explain that in the note but will check the documentation again.